src/analyses/base.ml

```diff
@@ -769,6 +769,102 @@ struct
           let a1 = eval_rv a gs st c1 in
           let a2 = eval_rv a gs st c2 in
           evalbinop a st op t1 a1 t2 a2 typ
+      | BinOp (LOr, arg1, arg2, typ) as exp ->
+        let (let*) = Option.bind in
+        (* split nested LOr Eqs to equality pairs, if possible *)
+        let rec split = function
+          (* copied from above to support pointer equalities with implicit casts inserted *)
+          | BinOp (op, (CastE (t1, e1) as c1), (CastE (t2, e2) as c2), typ) when typeSig t1 = typeSig t2 && (op = Eq || op = Ne) ->
+            let te1 = Cilfacade.typeOf e1 in
+            let te2 = Cilfacade.typeOf e2 in
+            let both_arith_type = isArithmeticType te1 && isArithmeticType te2 in
+            let is_safe = VD.is_safe_cast t1 te1 && VD.is_safe_cast t2 te2 && not both_arith_type in
+            M.tracel "cast" "remove cast on both sides for %a? -> %b\n" d_exp exp is_safe;
+            if is_safe then ( (* we can ignore the casts if the casts can't change the value *)
+              let e1 = if isArithmeticType te1 then c1 else e1 in
+              let e2 = if isArithmeticType te2 then c2 else e2 in
```

Comment:
Given that this is some non-trivial snippet, it would be nice to avoid code duplication and instead extracting a function for it.